storage/aws/antispam/aws.go

```diff
@@ -0,0 +1,437 @@
+		curIndex   uint64
+	)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+		}
+		size, err := lr.IntegratedSize(ctx)
+		if err != nil {
+			klog.Errorf("Populate: IntegratedSize(): %v", err)
+			continue
+		}
+
+		// Busy loop while there's work to be done
+		for workDone := true; workDone; {
+			err = func() error {
+				tx, err := f.as.dbPool.BeginTx(ctx, &sql.TxOptions{
+					ReadOnly: false,
+				})
+				if err != nil {
+					return err
+				}
+
+				row := tx.QueryRowContext(ctx, "SELECT nextIdx FROM AntispamFollowCoord WHERE id = 0")
+
+				var followFrom uint64
+				if err := row.Scan(&followFrom); err != nil {
+					return err
+				}
+
+				if followFrom >= size {
+					// Our view of the log is out of date, exit the busy loop and refresh it.
+					workDone = false
+					return nil
+				}
+
+				f.as.pushBack.Store(size-followFrom > uint64(f.as.opts.PushbackThreshold))
+
+				// If this is the first time around the loop we need to start the stream of entries now that we know where we want to
+				// start reading from:
+				if entryReader == nil {
+					next, st := lr.StreamEntries(ctx, followFrom)
+					stop = st
+					entryReader = newEntryStreamReader(next, f.bundleHasher)
+				}
+
+				bs := uint64(f.as.opts.MaxBatchSize)
+				if r := size - followFrom; r < bs {
+					bs = r
+				}
+				batch := make([][]byte, 0, bs)
+				for i := range int(bs) {
+					idx, c, err := entryReader.Next()
+					if err != nil {
+						return fmt.Errorf("entryReader.next: %v", err)
+					}
+					if wantIdx := followFrom + uint64(i); idx != wantIdx {
+						// We're out of sync
+						return errOutOfSync
+					}
+					batch = append(batch, c)
+				}
+				curEntries = batch
+				curIndex = followFrom
+
+				// Store antispam entries.
+				//
+				// Note that we're writing the antispam entries outside of the transaction here. The reason is because we absolutely do not want
+				// the transaction to fail if there's already an entry for the same hash in the IDSeq table.
+				//
+				// It looks unusual, but is ok because:
+				//  - individual antispam entries fails because there's already an entry for that hash is perfectly ok
+				//  - we'll only continue on to update FollowCoord if no errors (other than AlreadyExists) occur while inserting entries
+				//  - similarly, if we manage to insert antispam entries here, but then fail to update FollowCoord, we'll end up
+				//    retrying over the same set of log entries, and then ignoring the AlreadyExists which will occur.
+				{
+					sqlStr := "INSERT IGNORE INTO AntispamIDSeq (h, idx) VALUES "
+					vals := make([]any, 0, 2*len(curEntries))
+					for i, e := range curEntries {
+						sqlStr += "(?, ?),"
+						vals = append(vals, e, curIndex+uint64(i))
+					}
+					sqlStr = strings.TrimSuffix(sqlStr, ",")
+
+					_, err := f.as.dbPool.ExecContext(ctx, sqlStr, vals...)
+					if err != nil {
+						return fmt.Errorf("failed to insert into AntispamIDSeq with query %q: %v", sqlStr, err)
```

Comment:
Could potentially avoid the `TrimSuffix`:

```suggestion
					args := make([]string, 0, len(curEntries))
					vals := make([]any, 0, 2*len(curEntries))
					for i, e := range curEntries {
						args = append(args, "(?, ?)")
						vals = append(vals, e, curIndex+uint64(i))
					}
					sqlStr := fmt.Sprintf("INSERT IGNORE INTO AntispamIDSeq (h, idx) VALUES %s", strings.Join(args, ","))

					if _, err := f.as.dbPool.ExecContext(ctx, sqlStr, vals...); err != nil {
						return fmt.Errorf("failed to insert into AntispamIDSeq with query %q: %v", sqlStr, err)
```

Reply:
Yeah this is a bit tidier.